### PR TITLE
Security hardening and CI vulnerability scanning

### PR DESCRIPTION
**Pull request by mhenrixon**
## Summary

Full security audit of the pgbus codebase with fixes and safeguards:

- **CI: bundler-audit** — adds dependency vulnerability scanning to the CI pipeline (0 current vulnerabilities)
- **SQL injection prevention** — new `QueueNameValidator` module enforces strict alphanumeric+underscore validation on all queue names used in SQL identifiers, replacing scattered `gsub` calls
- **GlobalID type restriction** — new `config.allowed_global_id_models` allowlist prevents deserializing arbitrary objects from crafted event payloads (nil = allow all for backwards compat)
- **Dashboard security headers** — adds X-Frame-Options, X-Content-Type-Options, Referrer-Policy, Permissions-Policy
- **Dashboard auth warning** — logs a warning when `web_auth` is nil so users know the dashboard is open
- **Rails module safety** — fixes all `defined?(Rails)` checks to guard with `respond_to?` before calling methods, preventing `NoMethodError` when Rails module is partially loaded (e.g., by globalid without full framework)

### What was already safe
- CSRF protection, no XSS vectors, no Marshal/YAML.load, JSON-only serialization, thread-safe connection pooling, safe command validation in recurring jobs, no shell injection, no hardcoded secrets

## Test plan

- [ ] `bundle exec rspec spec/pgbus/queue_name_validator_spec.rb` — 12 examples for injection, length, blank names
- [ ] `bundle exec rspec spec/pgbus/security_spec.rb` — 8 examples for GlobalID validation, config validation, serialization safety
- [ ] `bundle exec rspec spec/pgbus/` — full suite passes (616 examples, 1 pre-existing timezone failure)
- [ ] `bundle exec rubocop` — no offenses
- [ ] `bundle exec bundle-audit check --update` — no vulnerabilities

## Summary by CodeRabbit

* **New Features**
  * CI adds automated dependency vulnerability scanning
  * Dashboard sets protective HTTP headers and warns when auth is unconfigured
  * Configurable allowlist for GlobalID model resolution; globalid declared as a dependency

* **Improvements**
  * Stricter, centralized queue-name validation and sanitization (alphanumeric + underscore, max length)
  * Safer GlobalID deserialization with explicit allowlist checks

* **Documentation**
  * CHANGELOG documents breaking change for restricted queue-name formats

* **Tests**
  * New specs for queue-name validation, security, dashboard auth, and GlobalID handling

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -33,5 +34,6 @@ group :development, :test do
   gem "actioncable", ">= 7.1", "< 9.0"
   gem "activejob", ">= 7.1", "< 9.0"
   gem "activerecord", ">= 7.1", "< 9.0"
+  gem "globalid", ">= 1.0"
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply Gemfile:37]: Fixed in 8a73e87. Moved `globalid` to `pgbus.gemspec` as a runtime dependency. Kept the Gemfile entry for dev/test as well since it pins the version for local development.

File: lib/pgbus/configuration.rb [hunk 2]
```diff
@@ -138,7 +139,9 @@ def initialize
     end
 
     def queue_name(name)
-      "#{queue_prefix}_#{name}"
+      full = "#{queue_prefix}_#{name}"
+      QueueNameValidator.validate!(full)
+      full
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/configuration.rb:144]: Fixed in 8a73e87. Added CHANGELOG entry documenting the breaking change with migration guidance. The dash restriction is intentional — PGMQ queue names become PostgreSQL table identifiers and dashes require quoting which PGMQ does not support.

File: lib/pgbus/recurring/config_loader.rb [hunk 3]
```diff
@@ -24,7 +24,7 @@ def load(path, env: nil)
       end
 
       def detect_env
-        if defined?(Rails)
+        if defined?(Rails) && Rails.respond_to?(:env)
           Rails.env.to_s
         else
           ENV.fetch("PGBUS_ENV", "development")
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/recurring/config_loader.rb]: Fixed in 8a73e87. Added `&& Rails.env` guard to `detect_env`. The parallel `config_loader.rb` already had this guard from the initial commit.

File: lib/pgbus/serializer.rb [hunk 4]
```diff
@@ -41,13 +41,30 @@ def deserialize_event(json_string)
...
+
+      allowed = Pgbus.configuration.allowed_global_id_models
+      if allowed&.none? { |klass| gid.model_class <= klass }
+        raise ArgumentError,
+              "GlobalID model #{gid.model_class} is not in allowed_global_id_models. " \
+              "Add it to Pgbus.configuration.allowed_global_id_models to permit deserialization."
+      end
+
+      GlobalID::Locator.locate(gid)
+    end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/serializer.rb:78]: Fixed in 8a73e87. Added explicit `empty?` check before the `none?` predicate with a clear error message: "GlobalID deserialization is disabled (allowed_global_id_models is empty)". Also added a test for this behavior.

File: lib/pgbus/web/authentication.rb [hunk 5]
```diff
@@ -13,12 +13,27 @@ module Authentication
...
+      def warn_unauthenticated_dashboard
+        return if @_pgbus_auth_warned
+
+        Pgbus.logger.warn do
+          "[Pgbus] Dashboard is accessible without authentication. " \
+            "Configure Pgbus.configuration.web_auth to restrict access. " \
+            "See: https://github.com/mhenrixon/pgbus#dashboard-authentication"
+        end
+        @_pgbus_auth_warned = true
+      end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/web/authentication.rb:40]: Fixed in 8a73e87. Moved the flag to a module-level `class << self; attr_accessor :auth_warned; end` on `Pgbus::Web::Authentication`. Now logs once per process instead of per request.

File: spec/pgbus/security_spec.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+      it "rejects models not in the allowlist" do
+        secret_class = Class.new
+        stub_const("Secret", secret_class)
+
+        expect do
+          Pgbus::Serializer.locate_global_id("gid://pgbus-test/Secret/1")
+        end.to raise_error(ArgumentError, /not in allowed_global_id_models/)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/pgbus/security_spec.rb:46]: Fixed in 8a73e87. Added test case `"when allowed_global_id_models is an empty array"` that verifies the explicit error message.

File: lib/pgbus/serializer.rb [hunk 7]
```diff
@@ -41,13 +41,35 @@ def deserialize_event(json_string)
...
+      if allowed&.empty?
+        raise ArgumentError,
+              "GlobalID deserialization is disabled (allowed_global_id_models is empty). " \
+              "Set to nil to allow all models, or add permitted classes."
+      end
+      if allowed&.none? { |klass| gid.model_class <= klass }
+        raise ArgumentError,
+              "GlobalID model #{gid.model_class} is not in allowed_global_id_models. " \
+              "Add it to Pgbus.configuration.allowed_global_id_models to permit deserialization."
+      end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/serializer.rb:75]: Fixed in b04b71c. Added type validation that checks each entry is a `Class` or `Module` before the `<=` comparison. Also allows modules for cases like `ActiveRecord::Base` ancestry checks. Added test for the string-instead-of-class edge case.

File: spec/pgbus/security_spec.rb [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      expect { config.queue_name("my;queue") }.to raise_error(ArgumentError)
+    end
+  end
+
+  describe "Dashboard authentication" do
+    it "defaults web_auth to nil" do
+      config = Pgbus::Configuration.new
+      expect(config.web_auth).to be_nil
+    end
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/pgbus/security_spec.rb:98]: Fixed in a191bac. Added spec in `authentication_spec.rb` that calls `authenticate_pgbus!` twice with `web_auth=nil` and asserts the logger warning is emitted exactly once (verifying the module-level flag works).
